### DISCUSSION TITLE
Type Error: React-Select compatibility issue on new version

### DISCUSSION THREAD

**Discussion by Noah-Mosher** [react-hook-form/react-hook-form#10605] [Q&A]
### Version Number

7.45.0

### Codesandbox/Expo snack

n/a

### Steps to reproduce

After updating to version 7.45.0, I am seeing this error on react-select select component

### Expected behaviour

The types should be compatible
  `          <div className={styles.selectContainer}>
              <Controller
                name={"region"}
                control={control}
                render={({ field }) => {
                  return (
                    <Select
                      className={styles.select}
                      {...field}
                      options={states}
                      placeholder="State"
                      styles={{
                        control: (baseStyles, state) => ({
                          ...baseStyles,
                          padding: "0 4px",
                          borderRadius: "8px",
                          borderColor: state.isFocused
                            ? "var(--ctaBlue)"
                            : "var(--gray12)",
                          borderWidth: "1px",
                          boxShadow: state.isFocused
                            ? "0px 0px 0px 4px var(--ctaBlue2)"
                            : "none",
                          height: "54px",
                          fontSize: "19px",
                          fontFamily: "var(--baseFont)",
                          color: "var(--gray02)",
                          zIndex: "10",
                        }),
                        menu: (provided) => ({ ...provided, zIndex: "9999" }),
                        menuList: (base) => ({
                          ...base,
                          "::-webkit-scrollbar": {
                            display: "none",
                          },
                          msOverflowStyle: "none",
                          fontFamily: "var(--baseFont)",
                          fontSize: "19px",
                          borderRadius: "8px",
                        }),
                        option: (provided, state) => ({
                          ...provided,
                          backgroundColor: state.isSelected
                            ? "var(--ctaBlue)"
                            : "white",
                        }),
                      }}
                      components={{
                        IndicatorSeparator: () => null,
                      }}
                    />
                  );
                }}
              />`

### Relevant log output

```shell
[{
	"resource": "/Users/nmosher/code/proteus/src/components/organisms/forms/AddressForm.tsx",
	"owner": "typescript",
	"code": "2322",
	"severity": 8,
	"message": "Type '{ options: { label: string; value: string; }[]; placeholder: string; styles: { control: (baseStyles: CSSObjectWithLabel, state: ControlProps<{ value: string; label: string; }, false, GroupBase<{ value: string; label: string; }>>) => { ...; }; menu: (provided: CSSObjectWithLabel) => { ...; }; menuList: (base: CSSObje...' is not assignable to type 'Partial<PublicBaseSelectProps<{ value: string; label: string; }, false, GroupBase<{ value: string; label: string; }>>>'.\n  Types of property 'onChange' are incompatible.\n    Type '(event: { value: string; label: string; } | ChangeEvent<Element>) => void' is not assignable to type '(newValue: SingleValue<{ value: string; label: string; }>, actionMeta: ActionMeta<{ value: string; label: string; }>) => void'.\n      Types of parameters 'event' and 'newValue' are incompatible.\n        Type 'SingleValue<{ value: string; label: string; }>' is not assignable to type '{ value: string; label: string; } | ChangeEvent<Element>'.\n          Type 'null' is not assignable to type '{ value: string; label: string; } | ChangeEvent<Element>'.",
	"source": "ts",
	"startLineNumber": 372,
	"startColumn": 22,
	"endLineNumber": 372,
	"endColumn": 28
}]
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by leapful** [comment]
@Noah-Mosher 

Please provide a codesandbox url to re-produce the issue. Only posting a part of your code implementation does not help to investigate the root cause.

**Comment by jonahallibone** [comment]
I am having the same issue.

I can reproduce here with `chakra-react-select`, which is just a wrapper for `react-select`

https://codesandbox.io/s/chakra-react-select-react-hook-form-usecontroller-typescript-forked-llkzr9?file=/app.tsx

I am getting:
```
Type '(event: string | boolean | FoodGroup | FoodGroup[] | ChangeEvent<Element> | undefined) => void' is not assignable to type '(newValue: unknown, actionMeta: ActionMeta<unknown>) => void'.
  Types of parameters 'event' and 'newValue' are incompatible.
    Type 'unknown' is not assignable to type 'string | boolean | FoodGroup | FoodGroup[] | ChangeEvent<Element> | undefined'
```

**Comment by leapful** [comment]
@jonahallibone 

I think the error is already self-described. The `onChange` prop from `react-select` component defines `newValue` as unknown type but `onChange` prop of `react-hook-form` expects defined type from your form values.

```
(newValue:IsMulti extends true ?ReadonlyArray<...>:SingleValue One of<...>, actionMeta:ActionMeta One of<...>) => void
```

You just need to cast it into correct path value of your form.

Working demo: https://codesandbox.io/s/chakra-react-select-react-hook-form-usecontroller-typescript-forked-h83ytm?file=/app.tsx

**Reply by jonahallibone** [reply]
So this kind of type strictness with onChange is something we should expect going forward?

**Reply by leapful** [reply]
@jonahallibone 

Yes, this is the expected behavior after type safe improvement on newer version of react-hook-form. Since you already know what type you expected for the generic component, it's safe for you to cast it to type for react-hook-form.
